README.md
# Шифр Цезаря
Приложение на Python, демонстрирующее [Шифр Цезаря](https://ru.wikipedia.org/wiki/%D0%A8%D0%B8%D1%84%D1%80_%D0%A6%D0%B5%D0%B7%D0%B0%D1%80%D1%8F) 

Реализованы 2 функции - шифрования и дешифрования. 
- Без использования специализированных библиотек и методов.
- С пропуском [суррогатных пар](https://en.wikipedia.org/wiki/UTF-16#U+D800_to_U+DFFF).
- С возможностью выбирать диапазон символов с которыми будем работать.
- Обе функции покрыты тестами на unittest.


CHANGELOG.md
# Change Log

## [Unreleased] - yyyy-mm-dd

### Changed
- Сделать одну общую функцию для шифрования и дешифрования чтобы следовать DRY принципу

### Added
- Добавить ограничение на длину сообщения для шифровки
- Добавить ограничение на величину смещения


## [0.3] - 2022-12-14

### Changed
- [(d53216d)](https://github.com/Pixadels/caesar_cipher/commit/d53216dadefa39341e4665cd056261fc1c36f29f) Улучшен визуал интерфейса ввода вывода

### Added
- [(2040319)](https://github.com/Pixadels/caesar_cipher/commit/20403190cf2b72ff36cf9251687b8a89beb5d667) Добавлена проверка введенных данных на соответствие типам 
- [(d53216d)](https://github.com/Pixadels/caesar_cipher/commit/d53216dadefa39341e4665cd056261fc1c36f29f) Добавлен бесконечный цикл на ввод вывод 
- [(aa826c4)](https://github.com/Pixadels/caesar_cipher/commit/aa826c46603880a1f2c412a5026df0c9d6b07480) Добавлены аннотации типов для new_str, new_ord, ALPHABET_LEN 


## [0.2] - 2022-12-14 

### Added [(fefec5b)](https://github.com/Pixadels/caesar_cipher/commit/fefec5b40bd8c390c1a462814d724bdfef237c0e)
- Добавлена обработка суррогатных пар в функции шифрования и дешифровки
- Добавлены новые тесты


## [0.1] - 2022-12-14 
 
### Added [(72a39b0)](https://github.com/Pixadels/caesar_cipher/commit/72a39b0daf62d6d23b8b51f8c226addce2f0a874)
- Реализован базовый функционал - шифрование и дешифровка
- Реализованы тесты для обоих функций


src/tests/test_main.py
import sys
sys.path.insert(0, 'src/main')

from main import encryptionCaesar
from main import decryptionCaesar
import unittest


class TestCeasar(unittest.TestCase):
    def test_encryption(self):
        """Тестируем шифровку.
        Если fork возвращает True, то тест успешно пройден
        """
        # обычная строка
        assert fork('abc', 1, 'bcd', True)
        # обычное число
        # self.assertEqual(encryptionCaesar('123', 1), '234')
        assert fork('abc', 1, 'bcd', True)
        # крайние левые символы
        assert fork(chr(0)+chr(1)+chr(2), 1, chr(0+1)+chr(1+1)+chr(2+1), True)
        # крайние правые символы
        assert fork(chr(0x10FFFF-2)+chr(0x10FFFF-1)+chr(0x10FFFF), 1,
                    chr(0x10FFFF-2+1)+chr(0x10FFFF-1+1)+chr(0), True)
        # 1 крайний левый символ
        assert fork(chr(0), 1, chr(0+1), True)
        # 1 крайний правый символ
        assert fork(chr(0x10FFFF), 1, chr(0), True)
        # нулевое смещение
        assert fork('abc', 0, 'abc', True)
        # отрицательное смещение
        assert fork('bcd', -1, 'abc', True)
        # обработка суррогатов
        assert fork(chr(0xD800-1), 1, chr(0xDFFF+1), True)

    def test_decryption(self):
        """Тестируем расшифровку.
        Если fork возвращает True, то тест успешно пройден
        """
        # обычная строка
        assert fork('abc', 1, 'bcd', False)
        # обычное число
        # self.assertEqual(encryptionCaesar('123', 1), '234')
        assert fork('abc', 1, 'bcd', False)
        # крайние левые символы
        assert fork(chr(0)+chr(1)+chr(2), 1, chr(0+1)+chr(1+1)+chr(2+1), False)
        # крайние правые символы
        assert fork(chr(0x10FFFF-2)+chr(0x10FFFF-1)+chr(0x10FFFF), 1,
                    chr(0x10FFFF-2+1)+chr(0x10FFFF-1+1)+chr(0), False)
        # 1 крайний левый символ
        assert fork(chr(0), 1, chr(0+1), False)
        # 1 крайний правый символ
        assert fork(chr(0x10FFFF), 1, chr(0), False)
        # нулевое смещение
        assert fork('abc', 0, 'abc', False)
        # отрицательное смещение
        assert fork('bcd', -1, 'abc', False)
        # обработка суррогатов
        assert fork(chr(0xD800-1), 1, chr(0xDFFF+1), False)
        # обработка ошибок
        # assert fork(1, 1, 2, False)


def fork(to_encode: str, shift: int, to_decode: str, encode: bool) -> bool:
    """Функция для упрощения написания тестов

    Принимает строку для кодирования, смещение, строку для декодирования,
    и режим работы encode.
    Аргумент encode позволяет быстро превратить набор тестов функции шифрования
    в набор тестов функции расшифровки и наоборот
    """
    if encode:
        if encryptionCaesar(to_encode, shift) == to_decode:
            return True
    else:
        if decryptionCaesar(to_decode, shift) == to_encode:
            return True
    return False


if __name__ == "__main__":
    unittest.main()


src/main/main.py
"""Модуль реализует шифр Цезаря.
На вход подаются текст для шифровки и смещение.
На выходе зашифрованная и расшифрованная строки.
"""

__version__ = '0.3'
__author__ = 'Linar Shayakhmetov'

ALPHABET_LEFT: int = 0x0  # Можно поменять левый край алфавита
ALPHABET_RIGHT: int = 0x10FFFF  # Можно поменять правый край алфавита
# Существуют служебные символы, которые я думаю будет красиво пропускать
# если новый символ будет попадать в этот диапазон
SUR_LEFT: int = 0xD800
SUR_RIGHT: int = 0xDFFF
SUR_LEN: int = SUR_RIGHT - SUR_LEFT + 1  # 57343 - 55296 + 1


def encryptionCaesar(text: str, shift: int) -> str:
    """Зашифруем полученную строку."""
    ALPHABET_LEN: int = ALPHABET_RIGHT - ALPHABET_LEFT + 1
    new_str: str = ''
    for ch in text:
        new_ord: int = ((ord(ch)+shift-ALPHABET_LEFT) % ALPHABET_LEN)
        new_ord += ALPHABET_LEFT
        # обработка суррогатов, они не принимаются и не возвращаются
        if SUR_LEFT <= new_ord <= SUR_RIGHT:
            new_str += chr(new_ord + SUR_LEN)
        else:
            new_str += chr(new_ord)
    return new_str


def decryptionCaesar(text: str, shift: int) -> str:
    """Расшифруем ранее зашифрованную строку."""
    shift *= -1
    ALPHABET_LEN: int = ALPHABET_RIGHT - ALPHABET_LEFT + 1
    new_str: str = ''
    for ch in text:
        new_ord: int = ((ord(ch)+shift-ALPHABET_LEFT) % ALPHABET_LEN)
        new_ord += ALPHABET_LEFT
        # обработка суррогатов, они не принимаются и не возвращаются
        if SUR_LEFT <= new_ord <= SUR_RIGHT:
            new_str += chr(new_ord - SUR_LEN)
        else:
            new_str += chr(new_ord)
    return new_str


def input_text() -> str:
    """Ввод сообщения для шифровки с обработкой ошибок"""
    try:
        text: str = str(input('Введите текст для шифровки: '))
        return text
    except ValueError:
        print("\nСообщение для шифровки должно быть строкой!")
        return input_text()


def input_shift() -> int:
    """Ввод смещения для шифровки с обработкой ошибок"""
    try:
        shift: int = int(input('Введите смещение: '))
        return shift
    except ValueError:
        print("\nСмещение должно быть целым числом!")
        return input_shift()


if __name__ == '__main__':
    while True:
        text: str = input_text()
        shift: int = input_shift()

        text_encrypt: str = encryptionCaesar(text, shift)
        text_decrypt: str = decryptionCaesar(text_encrypt, shift)

        print(f'\nРезультат шифровки: {text_encrypt}')
        print(f'Результат расшифровки: {text_decrypt}\n')
